pytorch_lightning/loops/closure.py

```diff
@@ -0,0 +1,132 @@
+    the :class:`ClosureResult` and caching it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._result: Optional[ClosureResult] = None
+
+    @property
+    def result(self) -> Optional[ClosureResult]:
+        """The cached result from the last time the closure was called. Once accessed, the internal reference
+        gets reset and the consumer will have to hold on to the reference as long as necessary."""
+        result = self._result
+        self._result = None  # free memory
+        return result
+
+    @abstractmethod
+    def closure(self, *args: Any, **kwargs: Any) -> Optional[ClosureResult]:
+        """Implements the behavior of the closure once it is getting called."""
+        pass
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
+        self._result = self.closure(*args, **kwargs)
+        if self._result is not None:
+            return self._result.loss
+
+
+class Closure(AbstractClosure):
+    """
+    An implementation of a :class:`AbstractClosure` for optimization in Lightning that combines three elementary
+    closures into one: ``training_step``, ``backward`` and ``zero_grad``.
+
+    The Closure gets created by the training loop(s) and is then passed to the
+    :meth:`torch.optim.Optimizer.step` method. An optimizer is responsible for calling the closure and optionally
+    do something with the output.
+
+    Args:
+        step_fn: This is typically the :meth:`pytorch_lightning.core.lightning.LightningModule.training_step
+            wrapped with processing for its outputs
+        backward_fn: A function that takes a loss value as input, performs back-propagation and returns the loss value.
+            Can be set to ``None`` to skip the backward operation.
+        zero_grad_fn: A function that zeroes the gradients. Can be set to ``None`` to skip zero_grad, for example
+            when accumulating gradients.
+        profiler: A profiler for profiling the actions of the passed in closure functions.
+
+    Example:
+
+        closure = Closure()
+        optimizer = torch.optim.Adam(...)
+        optimizer.step(closure)
+    """
+
+    def __init__(
+        self,
+        step_fn: Callable[[], dict],
+        backward_fn: Optional[Callable[[Tensor], Tensor]] = None,
+        zero_grad_fn: Optional[Callable[[], None]] = None,
+        profiler: Optional[BaseProfiler] = None,
+    ):
+        super().__init__()
+        self._step_fn = step_fn
+        self._backward_fn = backward_fn
+        self._zero_grad_fn = zero_grad_fn
+        self._profiler = PassThroughProfiler() if profiler is None else profiler
+
+    def closure(self, *args: Any, **kwargs: Any) -> Optional[ClosureResult]:
+        with self._profiler.profile("training_step_and_backward"):
+            output = self._step_fn()
+            output = ClosureResult(**output) if output else None
```

Comment:
Not sure, but I believe ResultCollection is always provided from step_fn, making `ClosureResult`never empty.

Reply:
the LightningModule training_step can return None to skip an iteration, so that's how we are handling it here